### ISSUE TITLE
[BUG][IPC4] IO error on TGLU RVP and Upextreme when play with SSP0 and SSP2  simultaneously

### ISSUE THREAD

**Issue by XiaoyunWu6666** [thesofproject/linux#3520]
**Describe the bug**
when play with SSP0 and SSP2  simultaneously , IO error happened 
FW error of  **FW reported error: 104 -Other failure of module instance initialization request**  happened first 
and **FW reported error: 9 - Specified resource not found** followed

noticed that fw error happened in inner test from 10699 March 03 to 10854 March 09
but I haven't observed fw error after March 09
**To Reproduce**
TPLG=/lib/firmware/intel/avs-tplg/cavs-tgl-nocodec.tplg MODEL=TGLU_RVP_NOCODEC_IPC4 ~/sof-test/test-case/multiple-pause-resume.sh -l 3 -r 5
OR
aplay -D hw:0,0 -r 48000 -c 2 -f S16_LE /dev/zero & aplay -D hw:0,2 -r 48000 -c 2 -f S16_LE /dev/zero

**Reproduction Rate**
almost 100% on TGLU_RVP_NOCODEC_IPC4
also reproducible on i11 UPextreme , Chromebook @ranj063 @bardliao 

**Environment**
Kernel Branch: integration/ipc4-wip
Kernel Commit: [fc948f43](https://github.com/thesofproject/linux/commit/fc948f43)
cavs fw
Topology:  /lib/firmware/intel/avs-tplg/sof-tgl-nocodec.tplg
Platform: TGLU

**Screenshots or console output**
[console]
```
~$ aplay -D hw:0,2 -r 48000 -c 2 -f S16_LE /dev/zero &  aplay -D hw:0,0 -r 48000 -c 2 -f S16_LE /dev/zero
[1] 2387
Playing raw data '/dev/zero' : Signed 16 bit Little Endian, Rate 48000 Hz, Stereo
Playing raw data '/dev/zero' : Signed 16 bit Little Endian, Rate 48000 Hz, Stereo
aplay: pcm_write:2061: write error: Input/output error

```
[dmesg]
```

[  412.488155] sof-audio-pci-intel-tgl 0000:00:1f.3: Create widget pipeline.1 instance 0 - pipe 1 - core 0
[  412.488160] sof-audio-pci-intel-tgl 0000:00:1f.3: pipeline: 1 memory pages: 1
[  412.488166] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx      : 0x11010001|0x0: GLB_CREATE_PIPELINE
[  412.488690] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx reply: 0x31000000|0x0: GLB_CREATE_PIPELINE
[  412.488716] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx done : 0x11010001|0x0: GLB_CREATE_PIPELINE
[  412.488720] sof-audio-pci-intel-tgl 0000:00:1f.3: widget pipeline.1 setup complete
[  412.488723] sof-audio-pci-intel-tgl 0000:00:1f.3: Create widget copier.host.1.1 instance 2 - pipe 1 - core 0
[  412.488730] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx      : 0x40020004|0x10015: MOD_INIT_INSTANCE [data size: 84]
[  412.489484] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx reply: 0x60000000|0x10015: MOD_INIT_INSTANCE
[  412.489508] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx done : 0x40020004|0x10015: MOD_INIT_INSTANCE [data size: 84]
[  412.489512] sof-audio-pci-intel-tgl 0000:00:1f.3: widget copier.host.1.1 setup complete
[  412.489516] sof-audio-pci-intel-tgl 0000:00:1f.3: Create widget pipeline.2 instance 0 - pipe 2 - core 0
[  412.489519] sof-audio-pci-intel-tgl 0000:00:1f.3: pipeline: 2 memory pages: 2
[  412.489523] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx      : 0x11020002|0x0: GLB_CREATE_PIPELINE
[  412.490365] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx reply: 0x31000000|0x0: GLB_CREATE_PIPELINE
[  412.490392] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx done : 0x11020002|0x0: GLB_CREATE_PIPELINE
[  412.490397] sof-audio-pci-intel-tgl 0000:00:1f.3: widget pipeline.2 setup complete
[  412.490399] sof-audio-pci-intel-tgl 0000:00:1f.3: Create widget copier.SSP.2.1 instance 3 - pipe 2 - core 0
[  412.490412] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx      : 0x40030004|0x2002d: MOD_INIT_INSTANCE [data size: 180]
[  412.490892] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx reply: 0x60000068|0x2002d: MOD_INIT_INSTANCE
[  412.490898] sof-audio-pci-intel-tgl 0000:00:1f.3: FW reported error: 104 - Other failure of module instance initialization request
[  412.490920] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc error for msg 0x40030004|0x2002d
[  412.490926] sof-audio-pci-intel-tgl 0000:00:1f.3: failed to create module copier.SSP.2.1
[  412.490933] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx      : 0x12020000|0x0: GLB_DELETE_PIPELINE
[  412.491410] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx reply: 0x32000000|0x0: GLB_DELETE_PIPELINE
[  412.491434] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx done : 0x12020000|0x0: GLB_DELETE_PIPELINE
[  412.491438] sof-audio-pci-intel-tgl 0000:00:1f.3: widget pipeline.2 instance 0 freed
[  412.491441] sof-audio-pci-intel-tgl 0000:00:1f.3: widget pipeline.2 freed
[  412.491445] sof-audio-pci-intel-tgl 0000:00:1f.3: widget copier.host.1.1 instance 2 freed
[  412.491449] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx      : 0x12010000|0x0: GLB_DELETE_PIPELINE
[  412.492086] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx reply: 0x32000000|0x0: GLB_DELETE_PIPELINE
[  412.492105] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx done : 0x12010000|0x0: GLB_DELETE_PIPELINE
[  412.492109] sof-audio-pci-intel-tgl 0000:00:1f.3: widget pipeline.1 instance 0 freed
[  412.492112] sof-audio-pci-intel-tgl 0000:00:1f.3: widget pipeline.1 freed
[  412.492114] sof-audio-pci-intel-tgl 0000:00:1f.3: widget copier.host.1.1 freed
[  412.492117] sof-audio-pci-intel-tgl 0000:00:1f.3: Failed to set up connected widgets
[  412.492779] sof-audio-pci-intel-tgl 0000:00:1f.3: pcm: trigger stream 0 dir 0 cmd 1
[  412.492779] sof-audio-pci-intel-tgl 0000:00:1f.3: pcm: trigger stream 2 dir 0 cmd 1
[  412.492818] sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0x160]=0x14001e successful
[  412.492824] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc4 set pipeline 5 state 3
[  412.492830] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx      : 0x13050003|0x0: GLB_SET_PIPELINE_STATE
[  412.493217] sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0x180]=0x24001e successful
[  412.493237] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc4 set pipeline 1 state 3
[  412.493369] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx reply: 0x33000000|0x0: GLB_SET_PIPELINE_STATE
[  412.493397] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx done : 0x13050003|0x0: GLB_SET_PIPELINE_STATE
[  412.493405] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc4 set pipeline 5 state 4
[  412.493409] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx      : 0x13010003|0x0: GLB_SET_PIPELINE_STATE
[  412.493788] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx reply: 0x33000009|0x1: GLB_SET_PIPELINE_STATE
[  412.493794] sof-audio-pci-intel-tgl 0000:00:1f.3: FW reported error: 9 - Specified resource not found
[  412.493813] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc error for msg 0x13010003|0x0
[  412.493822] sof-audio-pci-intel-tgl 0000:00:1f.3: failed to pause pipeline 1
[  412.493827] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx      : 0x13050004|0x0: GLB_SET_PIPELINE_STATE
[  412.493831] sof-audio-pci-intel-tgl 0000:00:1f.3: ASoC: error at soc_component_trigger on 0000:00:1f.3: -22
[  412.493842]  Port0: ASoC: trigger FE cmd: 1 failed: -22
[  412.493876] sof-audio-pci-intel-tgl 0000:00:1f.3: pcm: trigger stream 0 dir 0 cmd 0
[  412.493879] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc4 set pipeline 1 state 3
[  412.494361] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx reply: 0x33000000|0x0: GLB_SET_PIPELINE_STATE
[  412.494408] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx done : 0x13050004|0x0: GLB_SET_PIPELINE_STATE
[  412.494419] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc4 set pipeline 6 state 3
[  412.494426] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx      : 0x13010003|0x0: GLB_SET_PIPELINE_STATE
[  412.494645] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx reply: 0x33000009|0x1: GLB_SET_PIPELINE_STATE
[  412.494654] sof-audio-pci-intel-tgl 0000:00:1f.3: FW reported error: 9 - Specified resource not found
[  412.494679] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc error for msg 0x13010003|0x0
[  412.494689] sof-audio-pci-intel-tgl 0000:00:1f.3: failed to pause pipeline 1
[  412.494694] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx      : 0x13060003|0x0: GLB_SET_PIPELINE_STATE
[  412.494705] sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0x180]=0x240000 successful
[  412.494722] sof-audio-pci-intel-tgl 0000:00:1f.3: ASoC: error at soc_component_trigger on 0000:00:1f.3: -22
[  412.494727]  Port0: ASoC: trigger FE cmd: 0 failed: -22
[  412.495235] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx reply: 0x33000000|0x0: GLB_SET_PIPELINE_STATE
[  412.495294] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx done : 0x13060003|0x0: GLB_SET_PIPELINE_STATE
[  412.495301] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc4 set pipeline 6 state 4
[  412.495306] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx      : 0x13060004|0x0: GLB_SET_PIPELINE_STATE
[  412.495852] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx reply: 0x33000000|0x0: GLB_SET_PIPELINE_STATE
[  412.495871] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx done : 0x13060004|0x0: GLB_SET_PIPELINE_STATE
[  412.496392] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc rx      : 0x1b050000|0x0: GLB_NOTIFICATION|RESOURCE_EVENT
[  412.496461] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc rx done : 0x1b050000|0x0: GLB_NOTIFICATION|RESOURCE_EVENT [data size: 40]
[  412.496625] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc rx      : 0x1b050000|0x0: GLB_NOTIFICATION|RESOURCE_EVENT
[  412.496682] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc rx done : 0x1b050000|0x0: GLB_NOTIFICATION|RESOURCE_EVENT [data size: 40]
[  412.999876] sof-audio-pci-intel-tgl 0000:00:1f.3: pcm: free stream 0 dir 0
[  412.999886] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc4 set pipeline 1 state 3
[  412.999892] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx      : 0x13010003|0x0: GLB_SET_PIPELINE_STATE
[  413.000169] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx reply: 0x33000009|0x1: GLB_SET_PIPELINE_STATE
[  413.000176] sof-audio-pci-intel-tgl 0000:00:1f.3: FW reported error: 9 - Specified resource not found
[  413.000233] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc error for msg 0x13010003|0x0
[  413.000245] sof-audio-pci-intel-tgl 0000:00:1f.3: failed to pause pipeline 1
[  413.000312] sof-audio-pci-intel-tgl 0000:00:1f.3: ASoC: error at snd_soc_pcm_component_hw_free on 0000:00:1f.3: -22
[  413.000498] sof-audio-pci-intel-tgl 0000:00:1f.3: pcm: close stream 0 dir 0


```

**Comment by bardliao** [comment]
@XiaoyunWu6666 @ranj063 `multiple-pause-resume.sh -l 3 -r 5` and `run two aplay simultaneously` may not be the same issue. I can't see `FW reported error: ` when I run two aplay simultaneously.
[dmesg_ssp_simultaneously_playback.txt](https://github.com/thesofproject/sof/files/8211926/dmesg_ssp_simultaneously_playback.txt)

**Comment by XiaoyunWu6666** [comment]
@bard  you were trying on your i11 extreme?

I tried `multiple-pause-resume.sh -l 3 -r 5` and `run two aplay simultaneously` on TGLU_RVP_NOCODEC , both these two can cause the same error (I mean  FW error Other failure of module instance initialization request  first and Specified resource not found later)

Are we using the same recipe?

**Comment by bardliao** [comment]
> @bard you were trying on your i11 extreme?

Yes.

> 
> I tried `multiple-pause-resume.sh -l 3 -r 5` and `run two aplay simultaneously` on TGLU_RVP_NOCODEC , both these two can cause the same error (I mean FW error Other failure of module instance initialization request first and Specified resource not found later)
> 
> Are we using the same recipe?

Kernel: https://github.com/thesofproject/linux/commits/integration/ipc4-wip + https://github.com/thesofproject/linux/pull/3493. 
Firmware: closed-source firmware
Topology: Should be the same as yours as there is only one topology for TGL nocodec now.

**Comment by lgirdwood** [comment]
@RanderWang @marcinszkudlinski fyi.

**Comment by XiaoyunWu6666** [comment]
Sorry for misleading information , this failure is found with cavs fw + integration/ipc4-wip kernel , not sof.

**Comment by XiaoyunWu6666** [comment]
happened again in inner daily_IPC4 testresult  11165 on model TGLU_RVP_NOCODEC_IPC4 in testcase multiple-pause-resume-5

**Comment by juimonen** [comment]
@ranj063 did we get some clarity for this? I think the idea was that we are missing some crucial id in driver side and actually giving commands the same SSP instance...

**Comment by bardliao** [comment]
> @ranj063 did we get some clarity for this? I think the idea was that we are missing some crucial id in driver side and actually giving commands the same SSP instance...

@juimonen I saw 2 issues on my side.
1. We didn't set node index to node_id for SSP.
`copier_data->gtw_cfg.node_id |= SOF_IPC4_NODE_INDEX(ipc4_copier->dai_index);`
I am not sure if this should be done when we get blob. I.e. From intel_nhlt_get_endpoint_blob maybe.
2. After solving the node_id issue, I still got IO error when I start the 2nd playback.
This one can be "fixed" by increasing dma_buffer_size to 4 times.
```
-       copier_data->gtw_cfg.dma_buffer_size = available_fmt->dma_buffer_size[ret];
+       copier_data->gtw_cfg.dma_buffer_size = available_fmt->dma_buffer_size[ret] * 4;
```
I checked cavs_fw, all SSP modules share the same dma buffers.
My theory is that we calculate dma buffer size for a single stream, but it will not be enough when we start 2 streams simultaneously.
But I need to discuss the issue and solution with FW folks.
Any comment @ranj063  @RanderWang ?

**Comment by bardliao** [comment]
@ranj063 `copier_data->gtw_cfg.node_id |= SOF_IPC4_NODE_INDEX(ipc4_copier->dai_index);` for SSP dai is not quite right.
The struct of SSP Node index is
```
        uint8_t time_slot_group_index   : 4;
        uint8_t i2s_instance            : 4;
```

So the `ipc4_copier->dai_index` is actually `time_slot_group_index`, and that's why SSP0 and SSP1 shared the same DMA buffer in our test.
But if I try with `copier_data->gtw_cfg.node_id |= SOF_IPC4_NODE_INDEX(ipc4_copier->dai_index << 4);`, only SSP0 works on my Up Xtreme board.
i.e. I will see IO error whenever I play with SSP1 and SSP2.
I am suspecting that only SSP0 is supported on my Up Xtreme board. Could you check if SSP1 and SSP2 works on your I2S device with `copier_data->gtw_cfg.node_id |= SOF_IPC4_NODE_INDEX(ipc4_copier->dai_index << 4);`?

**Comment by plbossart** [comment]
> 2\. This one can be "fixed" by increasing dma_buffer_size to 4 times.

Bytes v. word confusion @bardliao ?

**Comment by ranj063** [comment]
> @ranj063 `copier_data->gtw_cfg.node_id |= SOF_IPC4_NODE_INDEX(ipc4_copier->dai_index);` for SSP dai is not quite right. The struct of SSP Node index is
> 
> ```
>         uint8_t time_slot_group_index   : 4;
>         uint8_t i2s_instance            : 4;
> ```
> 
> So the `ipc4_copier->dai_index` is actually `time_slot_group_index`, and that's why SSP0 and SSP1 shared the same DMA buffer in our test. But if I try with `copier_data->gtw_cfg.node_id |= SOF_IPC4_NODE_INDEX(ipc4_copier->dai_index << 4);`, only SSP0 works on my Up Xtreme board. i.e. I will see IO error whenever I play with SSP1 and SSP2. I am suspecting that only SSP0 is supported on my Up Xtreme board. Could you check if SSP1 and SSP2 works on your I2S device with `copier_data->gtw_cfg.node_id |= SOF_IPC4_NODE_INDEX(ipc4_copier->dai_index << 4);`?

@bardliao let me verify this change on my voxel

**Comment by ranj063** [comment]
> > 2. This one can be "fixed" by increasing dma_buffer_size to 4 times.
> 
> Bytes v. word confusion @bardliao ?

@plbossart the expected dma_buffer_size is in bytes and the value we're setting is also in bytes

**Comment by ranj063** [comment]
> So the `ipc4_copier->dai_index` is actually `time_slot_group_index`, and that's why SSP0 and SSP1 shared the same DMA buffer in our test.

@bardliao no this is incorrect. copier_data->gtw_cfg.node_id |= SOF_IPC4_NODE_INDEX(ipc4_copier->dai_index); is the right way to set the node_id. The FW assign the i2s_instance based on the node_id set above from the host

**Comment by ranj063** [comment]
> > So the `ipc4_copier->dai_index` is actually `time_slot_group_index`, and that's why SSP0 and SSP1 shared the same DMA buffer in our test.
> 
> @bardliao no this is incorrect. copier_data->gtw_cfg.node_id |= SOF_IPC4_NODE_INDEX(ipc4_copier->dai_index); is the right way to set the node_id. The FW assign the i2s_instance based on the node_id set above from the host

@bardliao you are right the node index is indeed bits 4-7 for SSP. After setting it correctly, the multiple pasuse-resume test passes for me

**Comment by RanderWang** [comment]
yes, it should be in bits 7 ~ 4. SOF FW has been fixed by COE team

**Comment by bardliao** [comment]
> > 2. This one can be "fixed" by increasing dma_buffer_size to 4 times.
> 
> Bytes v. word confusion @bardliao ?

@plbossart The node index will be translate to a i2s node index struct which is
```
    struct
    {
        uint8_t time_slot_group_index   : 4;
        uint8_t i2s_instance            : 4;
    }
```
The existing node index will map to the `time_slot_group_index` field, and `<< 4` is to map it to `i2s_instance`
See https://github.com/thesofproject/linux/pull/3548

**Comment by plbossart** [comment]
@ranj063 @XiaoyunWu6666 isn't this fixed according to test reports? can we close?

**Comment by ranj063** [comment]
fixed by #3548
